cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+  thrust::sort_by_key(handle.get_thrust_policy(),
+                      types_to_this_gpu.begin(),
+                      types_to_this_gpu.end(),
+                      freq_of_types_to_this_gpu.begin());
+
+  auto nr_unique_types_this_gpu =
+    thrust::count_if(handle.get_thrust_policy(),
+                     thrust::make_counting_iterator(size_t{0}),
+                     thrust::make_counting_iterator(types_to_this_gpu.size()),
+                     detail::is_first_in_run_t<edge_type_t const*>{types_to_this_gpu.data()});
+
+  rmm::device_uvector<edge_type_t> unique_types_to_this_gpu(nr_unique_types_this_gpu,
+                                                            handle.get_stream());
+  rmm::device_uvector<edge_t> freq_of_unique_types_this_gpu(nr_unique_types_this_gpu,
+                                                            handle.get_stream());
+
+  thrust::reduce_by_key(handle.get_thrust_policy(),
+                        types_to_this_gpu.begin(),
+                        types_to_this_gpu.end(),
+                        freq_of_types_to_this_gpu.begin(),
+                        unique_types_to_this_gpu.begin(),
+                        freq_of_unique_types_this_gpu.begin());
+
+  rmm::device_uvector<edge_type_t> global_unique_types(nr_unique_types_this_gpu,
+                                                       handle.get_stream());
+
+  thrust::transform(handle.get_thrust_policy(),
+                    unique_types_to_this_gpu.begin(),
+                    unique_types_to_this_gpu.end(),
+                    global_unique_types.begin(),
+                    [] __device__(auto val) { return int{val}; });
+
+  auto nr_unique_types_global = nr_unique_types_this_gpu;
+
+  if constexpr (multi_gpu) {
+    global_unique_types = cugraph::detail::shuffle_ext_vertices_to_local_gpu_by_vertex_partitioning(
+      handle, std::move(global_unique_types));
+
+    thrust::sort(
+      handle.get_thrust_policy(), global_unique_types.begin(), global_unique_types.end());
+
+    auto nr_unique_elements = static_cast<size_t>(thrust::distance(
+      global_unique_types.begin(),
+      thrust::unique(
+        handle.get_thrust_policy(), global_unique_types.begin(), global_unique_types.end())));
+
+    global_unique_types.resize(nr_unique_elements, handle.get_stream());
+
+    nr_unique_types_global = host_scalar_allreduce(
+      handle.get_comms(), nr_unique_elements, raft::comms::op_t::SUM, handle.get_stream());
+
+    global_unique_types = detail::device_allgatherv(
+      handle,
+      handle.get_comms(),
+      raft::device_span<int const>{global_unique_types.data(), global_unique_types.size()});
+
+    assert(global_unique_types.size() == nr_unique_types_global);
+  }
+
+  std::vector<edge_type_t> h_unique_types_global(nr_unique_types_global);
+
+  raft::update_host(h_unique_types_global.data(),
+                    global_unique_types.data(),
+                    global_unique_types.size(),
+                    handle.get_stream());
```

Comment:
So, the code up to this point is to find the number of types and edge IDs per type, right? And the number of types should be global and counts should be local (after shuffling based on hash_func(edge_id)), right? Let me think more about this. I think we can do this in a much cheaper way.